internal/opamp/opamp.go

```
@@ -0,0 +1,245 @@
+	if err != nil {
+		ctrl.logger.Error("Failed to unmarshal OpAMP AWSS3 Receiver custom message logs.", zap.Error(err))
+	}
+	return ctrl.DigForCompletionAndPublish(ctx, agentID, logMessage)
+}
+
+func (ctrl *OpAMPControlServer) DigForCompletionAndPublish(ctx context.Context, agentID string, logMessage plog.Logs) error {
+	foundCompletionLog := false
+	resourceLogs := logMessage.ResourceLogs()
+	for i := 0; i < resourceLogs.Len() && !foundCompletionLog; i++ {
+		resourceLog := resourceLogs.At(i)
+		scopeLogs := resourceLog.ScopeLogs()
+		for j := 0; i < scopeLogs.Len() && !foundCompletionLog; i++ {
+			scopeLog := scopeLogs.At(j)
+			logRecords := scopeLog.LogRecords()
+			rlen := logRecords.Len()
+			for k := range rlen {
+				logRecord := logRecords.At(k)
+				attributes := logRecord.Attributes()
+				if attribute, ok := attributes.Get(ingestStatusAttributeKey); ok {
+					statusAttrValue := attribute.AsString()
+					if statusAttrValue == ingestStatusCompleted || statusAttrValue == ingestStatusFailed {
+						agent := ctrl.agentUIDInfoMap[agentID]
+						if err := ctrl.PublishCompletionEvent(ctx, agentID, statusAttrValue); err != nil {
+							ctrl.logger.Error(
+								"Unable to publish replay completion event!",
+								zap.Error(err),
+								zap.String("instanceId", agent.instanceID),
+								zap.String("replayId", agent.replayID),
+								zap.String("replayOutcome", statusAttrValue),
+							)
+							return err
+						}
+						foundCompletionLog = true
+						break
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+type ReplayCompletionEventPayload struct {
+	ReplayID           string `json:"replay_id"`
+	ReplayResult       string `json:"replay_result"`
+	ReplayerInstanceID string `json:"replayer_instance_id"`
+}
+
+func (ctrl *OpAMPControlServer) PublishCompletionEvent(ctx context.Context, agentID string, outcome string) error {
+	agent := ctrl.agentUIDInfoMap[agentID]
```

should we fail fast if agent == nil here? also noticed you're using `agent.hubName` on the next line before you check for emptiness below it